posthog/models/feature_flag/flag_status.py

```diff
@@ -0,0 +1,173 @@
+import structlog
+
+from enum import StrEnum
+from .feature_flag import FeatureFlag
+
+logger = structlog.get_logger(__name__)
+
+FeatureFlagStatusReason = str
+
+
+class FeatureFlagStatus(StrEnum):
+    ACTIVE = "active"
+    INACTIVE = "inactive"
+    STALE = "stale"
+    DELETED = "deleted"
+    UNKNOWN = "unknown"
+
+
+# FeatureFlagStatusChecker is used to determine the status of a feature flag for a given user.
+# Eventually, this may be used to automatically archive old flags that are no longer in use.
+#
+# Status can be one of the following:
+# - ACTIVE: The feature flag is actively evaluated and the evaluations continue to vary.
+# - STALE: The feature flag has been fully rolled out to users. Its evaluations can not vary.
+# - INACTIVE: The feature flag is not being actively evaluated. STALE takes precedence over INACTIVE.
+# - DELETED: The feature flag has been soft deleted.
+# - UNKNOWN: The feature flag is not found in the database.
+class FeatureFlagStatusChecker:
+    def __init__(
+        self,
+        feature_flag_id: str,
+        # The amount of time considered "recent" for the purposes of determining staleness.
+        stale_window: str = "-30d",
+    ):
+        self.feature_flag_id = feature_flag_id
+        self.stale_window = stale_window
+
+    def get_status(self) -> tuple[FeatureFlagStatus, FeatureFlagStatusReason]:
+        try:
+            flag = FeatureFlag.objects.get(pk=self.feature_flag_id)
+        except FeatureFlag.DoesNotExist:
+            return FeatureFlagStatus.UNKNOWN, "Flag could not be found"
+
+        if flag.deleted:
+            return FeatureFlagStatus.DELETED, "Flag has been deleted"
+
+        # See if the flag is set to 100% on one variant (or 100% on rolled out and active if boolean flag).
+        is_flag_fully_enabled, fully_enabled_explanation = self.is_flag_fully_enabled(flag)
+        if is_flag_fully_enabled:
+            return FeatureFlagStatus.STALE, fully_enabled_explanation
+
+        # Final, and most expensive check: see if the flag has been evaluated recently.
+        if self.is_flag_unevaluated_recently(flag):
+            return FeatureFlagStatus.INACTIVE, "Flag has not been evaluated recently"
+
+        return FeatureFlagStatus.ACTIVE, "Flag is not fully rolled out and may still be active"
+
+    def is_flag_fully_enabled(self, flag: FeatureFlag) -> tuple[bool, FeatureFlagStatusReason]:
```

Comment:
I am worried about abstraction collisions between a flag being fully "enabled" (i.e. this flag is rolled out to everyone) vs "enabled" (i.e. "this flag is turned on", or "not disabled").  Feels like we have too many dimensions of status in here that I'd like to disambiguate a bit more somehow.

Reply:
Ah yup good point. "Rolled out" feels much more precise then. I've swapped out all references to "enabled" for "rolled out" in this file and the tests